Added async streams

ricardoalcaraz opened the pull request:
As promised here is the async stream. My use is to scrape entire subreddits. Sorry I had to make so many changes, some of them were necessary to ensure stability. Some notes is not to create an HttpClient and instead request from the DI container. And we always have to dispose the JsonDocument after we're done with it. My other changes were to conform to best practices. I tried my best to work with your existing implementation instead of stomping over it, let me know if you need any changes

ricardoalcaraz committed:
Added streaming scrolller services

ricardoalcaraz committed:
Removed unnecessary files

ricardoalcaraz committed:
updated namespaces

ricardoalcaraz committed:
updated test sdk

PereViader commented:
I think we can remove the old non streaming service and associated extractors. 
Having the whole json answer in POCOs  is much better.

PereViader commented on src/PereViader.Scrolller.Test/StreamingScrolllerServiceTests.cs:
```diff
@@ -0,0 +1,120 @@
+        catch (OperationCanceledException)
+        {
+            Assert.That(assets.Count, Is.EqualTo(MAX_URLS));
+        }
```
Looking at the code, I don't think that the exceptions in the tests will ever hit.
StreamingScrolller never calls `CancellationToken.ThrowIfCancellationRequested()`

ricardoalcaraz replied:
You're right we will not throw an exception however it does get thrown if we ever make another async request. I removed it because I realized we can just leave early

PereViader commented on src/PereViader.Scrolller.Test/StreamingScrolllerServiceTests.cs:
```diff
@@ -0,0 +1,120 @@
+    [TestCase("nsfw")]
+    public async Task ScrapeSubreddit_ShouldScrapeSomeUrls(string subreddit)
+    {
+        const int MAX_URLS = 100;
```
Could we make this a class level constant with a value of like 40-50 at most to not spam their endpoint.
And use the same constant for all the tests.

ricardoalcaraz replied:
Yeah I made it lower

ricardoalcaraz replied:
> Could we make this a class level constant with a value of like 40-50 at most to not spam their endpoint. And use the same constant for all the tests.

there's a way to intercept http calls and short circuit the response with a HttpResponseMessage of our own creation. We should do this for the unit tests and split actual api calls to integration tests if you want a proper solution

PereViader commented on src/PereViader.Scrolller/IStreamingScrolllerService.cs:
```diff
@@ -0,0 +1,11 @@
+    IAsyncEnumerable<IEnumerable<SubredditPostResponse>> Subreddit(string subreddit, CancellationToken ctx = default);
+    IAsyncEnumerable<IEnumerable<SubredditResponse>> Discover(bool isNsfw, CancellationToken ctx = default);
+    IAsyncEnumerable<Uri> DiscoverUrls(bool isNsfw, CancellationToken ctx = default);
+    IAsyncEnumerable<Uri> SubredditUrls(string subreddit, CancellationToken ctx = default);
```
Do you have a use case for these ones?
If not we could consider removing them from the library to keep it simple.

ricardoalcaraz replied:
> Do you have a use case for these ones? If not we could consider removing them from the library to keep it simple.

not really since it's easy enough to retrieve them from the full response. I removed this

ricardoalcaraz committed:
Simplified code and consolidated into one class

ricardoalcaraz commented:
> I think we can remove the old non streaming service and associated extractors. Having the whole json answer in POCOs is much better.

I kept the non streaming service since it still has it's use for one off cases. I changed everything to use POCOs and simplified stuff as a result.

PereViader commented:
Looks like tests have some issue.
Let's fix it, but other than that it looks good to me :)

PereViader commented:
@ricardoalcaraz No rush. Let me know if I can help with anything.

PereViader commented:
I'll take a look on why tests failed and merge it

PereViader commented:
https://github.com/ricardoalcaraz/Scrolller.Net/pull/1

Added these to your update so you can update this PR.
